Python/optimizer_symbols.c

```diff
@@ -48,187 +43,250 @@ static inline int get_lltrace(void) {
 {
-    // if the type version was already set, then it must be different and we should set it to bottom
-    if (sym->type_version) {
-        sym_set_bottom(ctx, sym);
-        return false;
+    JitSymType tag = sym->tag;
+    switch(tag) {
+        case JIT_SYM_NULL_TAG:
+            sym_set_bottom(ctx, sym);
+            return false;
+        case JIT_SYM_KNOWN_CLASS_TAG:
+            if (sym->cls.type->tp_version_tag != version) {
+                sym_set_bottom(ctx, sym);
+                return false;
+            }
+            else {
+                sym->cls.version = version;
+                return true;
+            }
+        case JIT_SYM_KNOWN_VALUE_TAG:
+            Py_CLEAR(sym->value.value);
+            sym_set_bottom(ctx, sym);
+            return false;
+        case JIT_SYM_TUPLE_TAG:
+            sym_set_bottom(ctx, sym);
+            return false;
+        case JIT_SYM_TYPE_VERSION_TAG:
+            if (sym->version.version == version) {
+                return true;
+            }
+            sym_set_bottom(ctx, sym);
+            return false;
+        case JIT_SYM_BOTTOM_TAG:
+            return false;
+        case JIT_SYM_NON_NULL_TAG:
+        case JIT_SYM_UNKNOWN_TAG:
+            sym->tag = JIT_SYM_TYPE_VERSION_TAG;
+            sym->version.version = version;
+            return true;
     }
-    sym->type_version = version;
-    return true;
+    Py_UNREACHABLE();
+}
+
+static void make_const(JitOptSymbol *sym, PyObject *val)
+{
+    sym->tag = JIT_SYM_KNOWN_VALUE_TAG;
+    sym->value.value = Py_NewRef(val);
 }
 
 void
-_Py_uop_sym_set_const(_Py_UOpsContext *ctx, _Py_UopsSymbol *sym, PyObject *const_val)
+_Py_uop_sym_set_const(JitOptContext *ctx, JitOptSymbol *sym, PyObject *const_val)
 {
-    assert(const_val != NULL);
-    if (sym->flags & IS_NULL) {
-        sym_set_bottom(ctx, sym);
-    }
-    PyTypeObject *typ = Py_TYPE(const_val);
-    if (sym->typ != NULL && sym->typ != typ) {
-        sym_set_bottom(ctx, sym);
-    }
-    if (sym->const_val != NULL) {
-        if (sym->const_val != const_val) {
-            // TODO: What if they're equal?
+    JitSymType tag = sym->tag;
+    switch(tag) {
+        case JIT_SYM_NULL_TAG:
             sym_set_bottom(ctx, sym);
-        }
-    }
-    else {
-        sym_set_flag(sym, NOT_NULL);
-        sym->typ = typ;
-        sym->const_val = Py_NewRef(const_val);
+            return;
+        case JIT_SYM_KNOWN_CLASS_TAG:
+            if (sym->cls.type != Py_TYPE(const_val)) {
+                sym_set_bottom(ctx, sym);
+                return;
+            }
+            make_const(sym, const_val);
+            return;
+        case JIT_SYM_KNOWN_VALUE_TAG:
+            if (sym->value.value != const_val) {
+                Py_CLEAR(sym->value.value);
+                sym_set_bottom(ctx, sym);
+            }
+            return;
+        case JIT_SYM_TUPLE_TAG:
+            sym_set_bottom(ctx, sym);
+            return;
```

Comment:
This also seems like it could potentially be fine (transitioning from an abstract tuple to a concrete one... could potentially need to do something recursive for the elements, though.